### PR TITLE
add create_zip and open_zip

### PR DESCRIPTION
**Pull request by samoconnor**
Per https://github.com/JuliaLang/METADATA.jl/pull/4332#issuecomment-168835129

### CODE REVIEW

**Comment by fhs** [comment]
This implementation is quiet different from your original proposal. You basically have a wrapper around Reader/Writer constructor and add some new methods to it. I think now it makes more sense to add the new functionality directly to Reader/Writer.

So, I propose renaming open_zip to Reader and create_zip to Writer, and bringing them back into top-level module.

That said, the following two function are kind of inconsistent:
```
Writer(Dict) -> data # no io or filename provided
Writer(io, Dict) -> nothing # io or filename provided
```
I'd change them to something like this:
```
encode(Dict) -> data # no io or filename provided
addfiles(io, Dict) -> nothing # io or filename provided
```
The name `addfiles` makes sense because we already have an `addfile` function. I'm not sure about `encode`. Maybe it should be `createzip` or `fromdict` or something else.

The documentation needs to be regenerated -- run doc/jldoc.py. But really we should be using docstrings introduced in julia 0.4.

File: src/create_zip.jl [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+create_zip(buf::Array{UInt8,1}) = create_zip(IOBuffer(buf, false, true))
+create_zip(f::Function, arg, args...) = with_close(f, create_zip(arg, args...))
+create_zip(filename::AbstractString) = Writer(filename)
+
+function create_zip(filename::AbstractString, arg, args...)
+    create_zip(open(filename, "w"), arg, args...)
+end
+
+
+# Use memory buffer if "filename" or "io" is provided.
```
Comment (fhs) [inline src/create_zip.jl]: is NOT provided

**Comment by samoconnor** [comment]
@fhs, My intention with this API design was to follow the patterns used in Base.
I hoped that this would make the API easy to understand for people who are familiar with Base.

My first thought was that Base should already have a good abstraction for reading and writing to an archive. Associative (and by implication iterable) seems like a good fit for an archive that identifies data by strings.

The next decision was type names. ZipFile has `Reader` and `Writer`, but "doer" data type names always seem wrong to me (Base only has 1 documented "doer": ClusterManager). `ZipArchive` seems like a better type name, but ZipFile.jl does not natively support updating zip archives, so it seemed best to stick with two types.

I agree that create_file() doing two different things isn't great (open file for writing vs one-shot archive creation).

In https://github.com/samoconnor/ZipFile.jl/commit/f275ea73c929063c8d8dd7687360c53e9fa4e955 I've refactored to avoid this. While I was at it I addressed the Reader/Writer issue and added a `ZipArchive` type that wraps Reader and Writer. `open_zip()` now returns ZipArchive. You can now add files, update files, read files, iterate over files etc all with one interface object. `create_zip()` is now reserved for one-shot creation of an archive from function arguments. See the commit log for full explanation.

Also new is this for creating a zip file from other files...
```julia
function create_zip(io::IO, files::Array)
    create_zip(io::IO, files, [open(readbytes, f) for f in files])
end

e.g.

create_zip("foo.zip", ["file1.csv", "file2.csv", "subdir/file3.csv"])
```
(prompted by a question from @nalimilan in https://github.com/JuliaLang/julia/pull/14546#issuecomment-169168864)

File: src/ZipFile.jl [hunk 2]
```diff
@@ -487,6 +500,9 @@ end
 
 # Write nb elements located at p into f.
 function write(f::WritableFile, p::Ptr, nb::Integer)
+    if nb == 0
+        return 0
+    end
```
Comment (samoconnor) [inline src/ZipFile.jl]: @fhs, bug fix for zero-length file written to archive.

File: src/ZipFile.jl [hunk 3]
```diff
@@ -352,7 +358,7 @@ function close(w::Writer)
 	_writele(w._io, @compat UInt32(cdpos))
 	_writele(w._io, @compat UInt16(0))
 
-	if !isa(w._io, IOBuffer)
+	if w._close_io
```
Comment (samoconnor) [inline src/ZipFile.jl]: @fhs, more general implementation of https://github.com/fhs/ZipFile.jl/pull/23

File: src/ZipFile.jl (hunk 2 above)
Comment (fhs) [inline src/ZipFile.jl]: I don't understand what's the bug. If nb is 0, `write` function wouldn't write anything anyway.

Comment (samoconnor) [inline src/ZipFile.jl]: it was crashing somewhere inside zlib when I wrote an empty file

File: src/ZipFile.jl (hunk 3 above)
Comment (fhs) [inline src/ZipFile.jl]: There is no need to add an extra flag. Add a `flush` function and call `flush` from from inside `close`. If a user don't want to close the underlying IOBuffer, they'll just call flush instead of close.

Please send this as a separate pull request. It's easier to review that way.

File: src/ZipFile.jl (hunk 2 above)
Comment (fhs) [inline src/ZipFile.jl]: It seems like a zlib or Zlib.jl bug. Maybe switching to Libz.jl will fix this (see #22).

For now, you can send a separate PR with a comment saying this is a workaround zlib/Zlib.jl bug.

**Comment by fhs** [comment]
I don't see why we need caching and ZipArchive. It seems like the library is trying to do too much. Caching is not an easy problem. You don't need want to read large files from a zip archive and keep it in memory.

I liked the previous implementation where it just added few more convenience functions to Reader/Writer. BTW, I called them Reader/Writer because this library was inspired by Go's zip package (https://golang.org/pkg/archive/zip/) and even in Go -er names are usually reserved for interfaces, not structs. So, I don't know if they're good names or not, but it's too late to change them.

Please discuss the new API before you implement it, so we can come to an agreement and you're not wasting your time ;-)

File: src/ZipFile.jl (hunk 3 above)
Comment (fhs) [inline src/ZipFile.jl]: I was wrong. Flush usually means to "clear the write buffer" but further writes should still be possible. However, if we write the central directory record when flushing, we won't be able to add more files to the zip file. We should be closing IO only if Reader/Writer owns the IO instance (i.e. we opened the file).

I'll take care of this one myself, so no worries.

File: src/ZipFile.jl (hunk 2 above)
Comment (fhs) [inline src/ZipFile.jl]: Never mind. I just merged the fix. Thanks.

**Comment by samoconnor** [comment]
Caching is necessary if you want to update an existing archive.
The library can either provide it, or leave it up to the user, but it can't be avoided.

A single ZipArchive type to replace Reader and Writer makes things much easier for the user who wants to read and write to the same archive.

My implementation is [careful to avoid the cache](https://github.com/samoconnor/ZipFile.jl/blob/master/src/open_zip.jl#L128) for use patterns where it is not needed, e.g. pure writes to a new archive, or pure reads from an existing archive.

The [test cases cover a number of combinations](https://github.com/samoconnor/ZipFile.jl/blob/master/test/runtests.jl#L269) of read-write-read, write-read-write, etc...

You are right that for large files there would be a large memory overhead.

I've done a quick implementation of the same interface using Info-ZIP as the backend.
This should be much more efficient for large files.
My guess is that the Info-ZIP implementation is faster in most cases, so I've made a seperate PR for it: See https://github.com/JuliaLang/METADATA.jl/pull/4382

I'll leave it up to you to decide how much of this interface is appropriate for ZipFile.

**Comment by samoconnor** [comment]
Hi @fhs, the `create_zip` and `open_zip` with Info-ZIP backend is now in a registered package over here: https://github.com/samoconnor/InfoZIP.jl

Although your pure-Julia ZipFile implementation is not as full featured as Info-ZIP I imagine that there are use cases where avoiding the external dependancy on Info-ZIP is desirable.

Do you have any interest in:
- merging the `create_zip` and `open_zip` interface in this PR into ZipFile, or
- me providing an option in the Info-ZIP package to use ZipFile as a backend in cases where Info-ZIP is not installed.

It seems to me that there is benefit to end users in having a consistent interface.

**Comment by fhs** [comment]
The latter option seems good to me. I would like to keep ZipFile.jl as pure julia.

I'm closing this PR for now. It seems to me that the use-case for this is dealing with lots of small files in a zip file. I think it'll introduce performance issues for large files in a zip file. Also, there are more important things to work on besides a higher-level alternate API, such as
* Better compression/depression performance (I looked at replacing Zlib.jl with Libz.jl but Libz.jl is missing stuff we need)
* ZIP64 support
* In general, support for more types of ZIP files, benchmarking and performance tuning.

**Comment by samoconnor** [comment]
Hi @fhs, sounds good to me.
See: https://github.com/samoconnor/InfoZIP.jl/commit/c1c1419914407b2a38187225a1f851b7a77b4888
